File: src/features/providers/builtin/openaiShared.ts
// ABOUTME: Shared OpenAI chat.completions / models wire helpers for TypeScript plugins.
// ABOUTME: Used by openai-compatible, deepseek, and related OpenAI-shaped plugins.
import {
  ProviderProtocolError,
  type ParsedModelPage,
  type ProviderHttpResponse,
  type ProviderWireRequest,
} from "../types";

const MAX_MODEL_KEY_LEN = 256;
const MAX_MODELS_PER_PAGE = 500;

export function normalizeModelKey(raw: string): string {
  const key = raw.trim();
  if (!key || key.length > MAX_MODEL_KEY_LEN) {
    throw new ProviderProtocolError("invalid model key");
  }
  return key;
}

export function parseOpenAiPage(response: ProviderHttpResponse): ParsedModelPage {
  let value: unknown;
  try {
    value = JSON.parse(response.body) as unknown;
  } catch {
    throw new ProviderProtocolError("model list is not JSON");
  }
  if (!value || typeof value !== "object") {
    throw new ProviderProtocolError("invalid model list page");
  }
  const data = (value as { data?: unknown }).data;
  if (!Array.isArray(data)) {
    throw new ProviderProtocolError("model list missing data array");
  }
  if (data.length > MAX_MODELS_PER_PAGE) {
    throw new ProviderProtocolError("model list page too large");
  }
  const items = data.map((entry) => {
    if (!entry || typeof entry !== "object") {
      throw new ProviderProtocolError("invalid model list entry");
    }
    const id = (entry as { id?: unknown }).id;
    if (typeof id !== "string") {
      throw new ProviderProtocolError("model list entry missing id");
    }
    return {
      modelKey: normalizeModelKey(id),
      remoteDisplayName: null,
      remoteMetadataJson: null,
    };
  });
  return { items, continuation: null };
}

export function openaiUserContent(userPrompt: string, imagePngBase64: string | null): unknown {
  if (imagePngBase64) {
    return [
      { type: "text", text: userPrompt },
      {
        type: "image_url",
        image_url: { url: `data:image/png;base64,${imagePngBase64}` },
      },
    ];
  }
  return userPrompt;
}

export function buildOpenAiChatCompletions(input: {
  modelKey: string;
  systemPrompt: string;
  userPrompt: string;
  temperature: number | null;
  maxTokens: number | null;
  imagePngBase64: string | null;
  stream: boolean;
}): ProviderWireRequest {
  const payload: Record<string, unknown> = {
    model: input.modelKey,
    messages: [
      { role: "system", content: input.systemPrompt },
      { role: "user", content: openaiUserContent(input.userPrompt, input.imagePngBase64) },
    ],
    stream: input.stream,
  };
  if (input.temperature != null) {
    payload.temperature = input.temperature;
  }
  if (input.maxTokens != null) {
    payload.max_tokens = input.maxTokens;
  }
  return {
    method: "POST",
    relativePath: "chat/completions",
    query: [],
    headers: { "content-type": "application/json" },
    body: JSON.stringify(payload),
  };
}

export function parseOpenAiChatContent(response: ProviderHttpResponse): string {
  let value: unknown;
  try {
    value = JSON.parse(response.body) as unknown;
  } catch {
    throw new ProviderProtocolError("chat response is not JSON");
  }
  if (!value || typeof value !== "object") {
    throw new ProviderProtocolError("invalid chat response");
  }
  const choices = (value as { choices?: unknown }).choices;
  if (!Array.isArray(choices) || choices.length === 0) {
    throw new ProviderProtocolError("chat response missing choices");
  }
  const message = (choices[0] as { message?: unknown }).message;
  if (!message || typeof message !== "object") {
    throw new ProviderProtocolError("chat response missing message");
  }
  const content = (message as { content?: unknown }).content;
  if (content == null) {
    throw new ProviderProtocolError("chat response missing content");
  }
  if (typeof content !== "string") {
    throw new ProviderProtocolError("chat content is not a string");
  }
  const trimmed = content.trim();
  if (!trimmed) {
    throw new ProviderProtocolError("chat content is empty");
  }
  return trimmed;
}

export function parseOpenAiStreamDelta(value: unknown): string | null {
  if (!value || typeof value !== "object") {
    return null;
  }
  const choices = (value as { choices?: unknown }).choices;
  if (!Array.isArray(choices) || choices.length === 0) {
    return null;
  }
  const delta = (choices[0] as { delta?: unknown }).delta;
  if (!delta || typeof delta !== "object") {
    return null;
  }
  const content = (delta as { content?: unknown }).content;
  if (typeof content !== "string" || content.length === 0) {
    return null;
  }
  return content;
}

File: src-tauri/src/adapters/builtin/openai_shared.rs
// ABOUTME: Shared OpenAI chat.completions / models wire helpers.
// ABOUTME: Used by openai-compatible, deepseek, and related OpenAI-shaped adapters.
use crate::adapters::transport::{ParsedPage, TransportError, build_endpoint};
use crate::domain::model::RemoteModelSyncItem;

const MAX_MODEL_KEY_LEN: usize = 256;
const MAX_MODELS_PER_PAGE: usize = 500;

pub fn normalize_model_key(raw: &str) -> Result<String, TransportError> {
  let key = raw.trim();
  if key.is_empty() || key.len() > MAX_MODEL_KEY_LEN {
    return Err(TransportError::InvalidResponse);
  }
  Ok(key.to_string())
}

/// Pure OpenAI models page parser: `{ data: [{ id }] }`.
pub fn parse_openai_page(value: &serde_json::Value) -> Result<ParsedPage, TransportError> {
  let data = value
    .get("data")
    .and_then(|v| v.as_array())
    .ok_or(TransportError::InvalidResponse)?;
  if data.len() > MAX_MODELS_PER_PAGE {
    return Err(TransportError::InvalidResponse);
  }
  let mut items = Vec::with_capacity(data.len());
  for entry in data {
    let id = entry
      .get("id")
      .and_then(|v| v.as_str())
      .ok_or(TransportError::InvalidResponse)?;
    let model_key = normalize_model_key(id)?;
    items.push(RemoteModelSyncItem {
      model_key,
      remote_display_name: None,
      remote_metadata_json: None,
      capability_overrides_json: None,
    });
  }
  Ok(ParsedPage {
    items,
    continuation: None,
  })
}

/// Build multimodal user content for OpenAI chat.completions when an image is present.
pub(crate) fn openai_user_content(user_prompt: &str, image_png_base64: Option<&str>) -> serde_json::Value {
  match image_png_base64 {
    Some(image) => serde_json::json!([
      { "type": "text", "text": user_prompt },
      {
        "type": "image_url",
        "image_url": {
          "url": format!("data:image/png;base64,{image}")
        }
      }
    ]),
    None => serde_json::json!(user_prompt),
  }
}

/// Apply OpenAI-compatible `thinking` control (`thinking.type` = enabled/disabled).
pub fn apply_openai_thinking(payload: &mut serde_json::Value, thinking: Option<bool>) {
  if let Some(enabled) = thinking {
    payload["thinking"] = serde_json::json!({
      "type": if enabled { "enabled" } else { "disabled" }
    });
  }
}

/// Build OpenAI chat.completions URL + body (stream or non-stream).
pub(crate) fn build_openai_chat_completions(
  base_url: &str,
  model_key: &str,
  system_prompt: &str,
  user_prompt: &str,
  temperature: Option<f64>,
  max_tokens: Option<u32>,
  thinking: Option<bool>,
  image_png_base64: Option<&str>,
  stream: bool,
) -> Result<(url::Url, serde_json::Value), TransportError> {
  let url = build_endpoint(base_url, "chat/completions")?;
  let mut payload = serde_json::json!({
    "model": model_key,
    "messages": [
      { "role": "system", "content": system_prompt },
      { "role": "user", "content": openai_user_content(user_prompt, image_png_base64) }
    ],
    "stream": stream
  });
  if let Some(temp) = temperature {
    payload["temperature"] = serde_json::json!(temp);
  }
  if let Some(max) = max_tokens {
    payload["max_tokens"] = serde_json::json!(max);
  }
  apply_openai_thinking(&mut payload, thinking);
  Ok((url, payload))
}

/// Final answer is always `message.content`. Thinking providers may also populate
/// sibling `reasoning_content` with chain-of-thought; that is never the answer.
pub fn parse_openai_chat_content(value: &serde_json::Value) -> Result<String, TransportError> {
  let message = value
    .get("choices")
    .and_then(|c| c.as_array())
    .and_then(|arr| arr.first())
    .and_then(|choice| choice.get("message"))
    .ok_or(TransportError::InvalidResponse)?;
  let content = match message.get("content") {
    None => return Err(TransportError::InvalidResponse),
    // Null content is treated like empty — common while only reasoning was produced.
    Some(serde_json::Value::Null) => "",
    Some(v) => v.as_str().ok_or(TransportError::InvalidResponse)?,
  };
  let trimmed = content.trim();
  if trimmed.is_empty() {
    // Complete non-stream response with no final answer (e.g. max_tokens spent on
    // reasoning_content). Not a mid-stream wait state — nothing more will arrive.
    return Err(TransportError::InvalidResponse);
  }
  Ok(trimmed.to_string())
}

/// OpenAI chat.completions stream chunk: `choices[0].delta.content`.
///
/// Thinking-mode providers stream CoT first via `delta.reasoning_content` while
/// `content` is null, absent, or empty. Those chunks are skipped so the consumer
/// keeps waiting for later final-answer deltas.
pub fn parse_openai_stream_delta(value: &serde_json::Value) -> Option<String> {
  let delta = value
    .get("choices")
    .and_then(|c| c.as_array())
    .and_then(|arr| arr.first())
    .and_then(|choice| choice.get("delta"))?;
  let content = match delta.get("content") {
    Some(serde_json::Value::String(s)) => s.as_str(),
    // null / missing / non-string → wait for subsequent chunks
    _ => return None,
  };
  if content.is_empty() {
    None
  } else {
    Some(content.to_string())
  }
}
